tutorials/sphinx-tutorials/beam_search_with_gpt.py

```diff
@@ -0,0 +1,414 @@
+# associated (observation, action) pair, we use a (hash, action) pair. This approach has multiple advantages:
+#
+# - Observations have a variable shape, making it hard to preallocate storage or store them contiguously and efficiently
+#   in memory.
+# - Using observations directly incurs extra memory cost as duplicated data will be stored (since successive steps in a
+#   branch share several tokens). Successive nodes only differ by an extra action (i.e., an extra token), and
+#   lateral (sibling) nodes differ only by their action.
+#   The only information we need to store is the action associated with the node itself. To reconstruct the sequence of
+#   tokens up to a certain node, we concatenate the actions along the path to that node - an operation for which
+#   torchrl has all the necessary tooling.
+#
+# The environment has two main methods: `_reset`, which initializes the environment with a given observation, and
+# `_step`, which takes an action (i.e., the next token in the sequence) and returns the updated observation.
+#
+# .. figure:: /_static/img/rollout-llm.png
+#    :alt: Data collection loop with our LLM environment.
+#
+
+env = LLMHashingEnv(vocab_size=tokenizer.vocab_size, tokenizer=tokenizer)
+
+################################################
+# Define the policy
+# -----------------
+#
+# In this section, we define a "policy" that takes an observation as input and outputs an action. Note that in our
+# context, the term "policy" is used to fit into control frameworks, but at its core, it's simply a language model
+# with some additional pre and post-processing steps.
+#
+# Policy Architecture
+# ~~~~~~~~~~~~~~~~~~~
+#
+# Our policy consists of a sequence of modules
+#
+# 1. Select unique observations (or nodes) in the input data.
+# 2. LLMWrapper: This module wraps the GPT-2 model and provides a convenient interface for generating output.
+# 3. Select last logit: This module selects the last logit from the output of the LLMWrapper.
+# 4. Probabilistic sampling: This module samples from a categorical distribution to select the next token.
+# 5. Reshape: This module reshapes the output to a 1D tensor.
+# 6. Top-k selection: This module selects the top-k tokens with the highest probabilities.
+#
+# Selecting unique observations
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# It might be the case that the policy receives multiple identical states in the batch.
+# To make the computation more efficient, we want to select only the unique values of these observations.
+# Doing this also reduces the chances that we will be generating redundant trajectories.
+# Because we have hashes that uniquely define the trajectory up to a given step, it's easier to use these hash values
+# to pinpoint the unique nodes rather than using the observations directly.
+#
+# Notice that indexing the relevant nodes is made easy thanks to tensordict's API!
+#
+
+
+def select_unique_obs(td):
+    # Get the obs (the hash)
+    hashes = td["hash"]
```

Comment:
Looks like `LLMHashingEnv` currently uses the key "hashing", not "hash", so this currently fails in the generated tutorial: https://docs-preview.pytorch.org/pytorch/rl/2623/tutorials/beam_search_with_gpt.html#run-policy

Reply:
Good point!
I think we should avoid any "hash" that refers to a hashing value, bc it may override a built-in in some cases (eg `hash = do_smth()`)